File: test/find.cpp
//
// Copyright (C) 2022 by Oliver John Hitchcock - github.com/c0rp3n
//
// Distributed under the Boost Software License, Version 1.0.
//

#define BOOST_UT_DISABLE_MODULE
#include <boost/ut.hpp>

#include "cento/cento.hpp"
#include "cento/centoFind.hpp"

using namespace boost::ut;

suite find = []()
{
    "universe"_test = []()
    {
        cento::Plane plane;

        const cento::Tile* const t = cento::createUniverse(plane);

        expect(cento::findTileAt(plane, {0,0})    == t);
        expect(cento::findTileAt(plane, {-512,0}) == t);
        expect(cento::findTileAt(plane, {512,0})  == t);
        expect(cento::findTileAt(plane, {0,-512}) == t);
        expect(cento::findTileAt(plane, {0,512})  == t);
    };

    "axis"_test = []()
    {
        cento::Plane plane;

        /*
         *         :
         *         :
         *    tl   :   tr
         *         :
         *         :
         * - - - - + - - - -
         *         :
         *         :
         *    bl   :   br
         *         :
         *         :
         *
         */

        constexpr const i32 min = cento::nInfinity;
        constexpr const i32 max = cento::pInfinity;

        const cento::TilePlan tlp{.id   = 0,
                                  .rect = {.ll = {.x = min, .y = 0},
                                           .ur = {.x = 0, .y = max}}};

        const cento::TilePlan trp{.id   = 1,
                                  .rect = {.ll = {.x = 0, .y = 0},
                                           .ur = {.x = max, .y = max}}};

        const cento::TilePlan blp{.id   = 2,
                                  .rect = {.ll = {.x = min, .y = min},
                                           .ur = {.x = 0, .y = 0}}};

        const cento::TilePlan brp{.id   = 3,
                                  .rect = {.ll = {.x = 0, .y = min},
                                           .ur = {.x = max, .y = 0}}};

        cento::Tile* const tl = cento::createTile(plane, tlp);
        cento::Tile* const tr = cento::createTile(plane, trp);
        cento::Tile* const bl = cento::createTile(plane, blp);
        cento::Tile* const br = cento::createTile(plane, brp);

        tl->below = bl;
        tl->right = tr;

        tr->below = br;
        tr->left  = tl;

        bl->above = tl;
        bl->right = br;

        br->above = tr;
        br->left  = bl;

        plane.start = bl;

        expect(cento::findTileAt(plane, {0,0})       == bl);
        expect(cento::findTileAt(plane, {-256,256})  == tl);
        expect(cento::findTileAt(plane, {256,256})   == tr);
        expect(cento::findTileAt(plane, {-256,-256}) == bl);
        expect(cento::findTileAt(plane, {256,-256})  == br);
    };
};

File: cento/tests/find.rs
use cento::*;
use geo::coord;

#[test]
fn universe() {
    let plane = Plane::new();
    let t = plane.hint();

    assert_eq!(plane.find_tile_at(Point::new(0, 0)), t);
    assert_eq!(plane.find_tile_at(Point::new(-512, 0)), t);
    assert_eq!(plane.find_tile_at(Point::new(512, 0)), t);
    assert_eq!(plane.find_tile_at(Point::new(0, -512)), t);
    assert_eq!(plane.find_tile_at(Point::new(0, 512)), t);
}

#[test]
fn axis() {
    //
    //         :
    //         :
    //    tl   :   tr
    //         :
    //         :
    // - - - - + - - - -
    //         :
    //         :
    //    bl   :   br
    //         :
    //         :
    //

    let tiles = vec![
        Tile::new(
            Rect::new(coord! { x: i32::MIN, y: 0}, coord! { x: 0, y: i32::MAX}),
            None,
        ),
        Tile::new(
            Rect::new(coord! { x: 0, y: 0}, coord! { x: i32::MAX, y: i32::MAX}),
            None,
        ),
        Tile::new(
            Rect::new(coord! { x: i32::MIN, y: i32::MIN}, coord! { x: 0, y: 0}),
            None,
        ),
        Tile::new(
            Rect::new(coord! { x: 0, y: i32::MIN}, coord! { x: i32::MAX, y: 0}),
            None,
        ),
    ];

    let (mut plane, keys) = unsafe { Plane::from_unchecked(tiles) };
    assert_eq!(keys.len(), 4);

    let tl = keys[0];
    let tr = keys[1];
    let bl = keys[2];
    let br = keys[3];

    below_mut!(plane, tl) = Some(bl);
    right_mut!(plane, tl) = Some(tr);

    below_mut!(plane, tr) = Some(br);
    left_mut!(plane, tr) = Some(tl);

    above_mut!(plane, bl) = Some(tl);
    right_mut!(plane, bl) = Some(br);

    above_mut!(plane, br) = Some(tr);
    left_mut!(plane, br) = Some(bl);

    assert_eq!(plane.find_tile_at(Point::new(0, 0)), tr);
    assert_eq!(plane.find_tile_at(Point::new(-256, 256)), tl);
    assert_eq!(plane.find_tile_at(Point::new(256, 256)), tr);
    assert_eq!(plane.find_tile_at(Point::new(-256, -256)), bl);
    assert_eq!(plane.find_tile_at(Point::new(256, -256)), br);
}
